core/src/main/java/com/linecorp/armeria/server/healthcheck/ScheduledHealthChecker.java

```diff
@@ -0,0 +1,205 @@
+        }
+        return true;
+    }
+
+    /**
+     * Health checker can be scheduled only once. Calling startHealthChecker won't work after stopHealthChecker.
+     */
+    private static class ScheduledHealthCheckerImpl implements ListenableHealthChecker {
+
+        private final Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker;
+        private final Duration maxTtl;
+        private final SettableHealthChecker settableHealthChecker;
+        private final EventExecutor eventExecutor;
+
+        private State state;
+        private volatile Future<?> scheduledFuture;
+
+        ScheduledHealthCheckerImpl(Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker,
+                                   Duration maxTtl, EventExecutor eventExecutor) {
+            this.healthChecker = healthChecker;
+            this.maxTtl = maxTtl;
+            this.eventExecutor = eventExecutor;
+            settableHealthChecker = new SettableHealthChecker(false);
+
+            state = State.INIT;
+            scheduledFuture = Futures.immediateVoidFuture();
+        }
+
+        @Override
+        public boolean isHealthy() {
+            return settableHealthChecker.isHealthy();
+        }
+
+        @Override
+        public void addListener(Consumer<? super HealthChecker> listener) {
+            settableHealthChecker.addListener(listener);
+        }
+
+        @Override
+        public void removeListener(Consumer<?> listener) {
+            settableHealthChecker.removeListener(listener);
+        }
+
+        private void startHealthChecker() {
+            if (state != State.INIT) {
+                return;
+            }
+            state = State.SCHEDULED;
+            eventExecutor.execute(this::runHealthCheck);
+        }
+
+        private void stopHealthChecker() {
+            if (state != State.SCHEDULED) {
+                return;
+            }
+            state = State.FINISHED;
+            scheduledFuture.cancel(true);
+        }
+
+        private boolean isActive() {
+            return state == State.SCHEDULED;
+        }
+
+        private void runHealthCheck() {
+            if (state != State.SCHEDULED) {
+                return;
+            }
+            try {
+                healthChecker.get().handle((result, throwable) -> {
+                    final boolean isHealthy;
+                    final long intervalMills;
+                    if (throwable != null) {
+                        isHealthy = false;
+                        intervalMills = maxTtl.toMillis();
+                    } else {
+                        isHealthy = result.isHealthy();
+                        intervalMills = result.ttlMillis();
+                    }
+                    settableHealthChecker.setHealthy(isHealthy);
+                    scheduledFuture = eventExecutor.schedule(this::runHealthCheck, intervalMills,
+                                                             TimeUnit.MILLISECONDS);
```

Comment:
How about not using `.schedule()` at all? `ScheduledHealthChecker` could do something like this instead:

```java
final AtomicReference<CompletableFuture<HealthCheckStatus>> currentFuture = new AtomicReference<>();
volatile long nextCheckTimeNanos = System.currentTimeNanos();

public boolean isHealthy() {
    for (;;) {
        CompletableFuture<HealthCheckStatus> oldFuture = currentFuture.get();
        if (!oldFuture.isDone() || System.nanoTime() - nextCheckTimeNanos < 0) {
            return lastKnownHealthiness;
        }

        CompletableFuture<HealthCheckStatus> newFuture = new CompletableFuture<>();
        if (currentFuture.compareAndSet(oldFuture, newFuture)) {
            doHealthCheckOnce(newFuture);
        }

        return lastKnownHealthiness;
    }
}

void start() {
    numUsers++;
}

void stop() {
    if (--numUsers <= 0) {
        CompletableFuture<HealthCheckStatus> oldFuture = currentFuture.get()'
        if (oldFuture != null) {
            oldFuture.cancel();
        }
    }
}
```

Reply:
If using isHealthy to trigger, maybe we don't need to count number of users. Because it's ok to wait final check complete even it becomes 0 user.

I think solution is simpler and it works well for high scale server. But just think if all the client is using long polling, small number of clients, if long polling time larger than user expect `delay` of health checker as above concern.(Currently only reconnected healtch check will trigger isHealthy again, if I checked code correctly) e.g. 3 clients with 1min long polling and 10 seconds cycle HealthChecker may have around 1min periodic check for extreme case.

Reply:
How about calling isHealthy() every few seconds for all ScheduledHealthCheckers in the background from HealthCheckService while the server is running? 🤔 

At second thought, I think it may not be worth doing what I suggested, then. Let's leave it as it is for now. Let me let you know if I have a better idea.